packages/relay/src/lib/config/preconfiguredSpendingPlans.ts

```diff
@@ -0,0 +1,127 @@
+  }
+  try {
+    const rawData = fs.readFileSync(configPath, 'utf-8');
+    return JSON.parse(rawData) as SpendingPlanConfig[];
+  } catch (error: any) {
+    throw new Error(`Failed to parse JSON from ${configPath}: ${error.message}`);
+  }
+};
+
+/**
+ * Validates the spending plan configuration.
+ * @param {SpendingPlanConfig[]} spendingPlans - The spending plan configurations to validate.
+ * @throws {Error} If any spending plan configuration is invalid.
+ */
+const validateSpendingPlanConfig = (spendingPlans: SpendingPlanConfig[]): void => {
+  for (const plan of spendingPlans) {
+    if (!isValidSpendingPlanConfig(plan)) {
+      throw new Error(`Invalid spending plan configuration: ${JSON.stringify(plan)}`);
+    }
+  }
+};
+
+/**
+ * Populates the database with pre-configured spending plans.
+ * @param {Logger} logger - The logger instance.
+ * @param {HbarSpendingPlanRepository} hbarSpendingPlanRepository - The HBAR spending plan repository.
+ * @param {EthAddressHbarSpendingPlanRepository} ethAddressHbarSpendingPlanRepository - The ETH address HBAR spending plan repository.
+ * @param {IPAddressHbarSpendingPlanRepository} ipAddressHbarSpendingPlanRepository - The IP address HBAR spending plan repository.
+ * @returns {Promise<void>} - A promise that resolves when the operation is complete.
+ * @throws {Error} - If the spending plans configuration file is not found or cannot be loaded.
+ * @async
+ */
+export const populatePreconfiguredSpendingPlans = async (
+  logger: Logger,
+  hbarSpendingPlanRepository: HbarSpendingPlanRepository,
+  ethAddressHbarSpendingPlanRepository: EthAddressHbarSpendingPlanRepository,
+  ipAddressHbarSpendingPlanRepository: IPAddressHbarSpendingPlanRepository,
+): Promise<void> => {
+  const requestDetails = new RequestDetails({ requestId: '', ipAddress: '' });
+  const ttl = PRE_CONFIGURED_SPENDING_PLANS_TTL;
+
+  const spendingPlans = loadSpendingPlansConfig(process.env.HBAR_SPENDING_PLANS_CONFIG_FILE);
+  validateSpendingPlanConfig(spendingPlans);
+
+  for (const plan of spendingPlans) {
+    const { name, ethAddresses, ipAddresses, subscriptionTier } = plan;
+
+    if (!ethAddresses?.length && !ipAddresses?.length) {
+      logger.warn(
+        `${requestDetails.formattedRequestId} Skipping HBAR spending plan '${name}' as it has no ethAddresses or ipAddresses`,
+      );
+      continue;
+    }
+
+    const spendingPlan = await hbarSpendingPlanRepository.create(subscriptionTier, requestDetails, ttl);
+    logger.info(`${requestDetails.formattedRequestId} Created HBAR spending plan '${name}' with ID ${spendingPlan.id}`);
+
+    if (ethAddresses) {
+      for (const ethAddress of ethAddresses) {
+        if (!(await ethAddressHbarSpendingPlanRepository.existsByAddress(ethAddress, requestDetails))) {
+          await ethAddressHbarSpendingPlanRepository.save({ ethAddress, planId: spendingPlan.id }, requestDetails, ttl);
```

Comment:
Q: So, this means that each spending plan can be associated with multiple `ethAddresses`. For instance, a plan with `planID apple-36` is associated with the addresses `0x33`, `0x66`, and `127.0.0.1`. When these addresses make requests to the relay, all expenses from these addresses are recorded and attributed to the `apple-36` plan, correct?

Secondly, I noticed that the `spendingPlan: IDetailedHbarSpendingPlan` only tracks `amountSpent` and `spendingHistory`, but neither of these seems to reference the addresses causing the expenses. Just thinking out loud here—wouldn't it be beneficial if the `IDetailedHbarSpendingPlan` could track how much each address has spent? From the current implementation, if in any cases that the Relay wants to know how much an individual address has spent, it would have to query the cache, correct? Now including the addresses in the usage history of the plan could remove the need to query the cache and potentially offer some benefits for metrics.

Reply:
Yes, this is the way it is supposed to work.

Btw, we have disabled tracking of spending history at the moment (as discussed in the daily a week ago) because we don't know yet how we want to use it, but we could include this information in the `HbarSpendingRecords` if we want to track this:
```diff
export class HbarSpendingRecord implements IHbarSpendingRecord {
  amount: number;
  timestamp: Date;
+  ethAddress?: string;
+  ipAddress?: string;

  constructor(data: IHbarSpendingRecord) {
    this.amount = data.amount;
    this.timestamp = new Date(data.timestamp);
+    this.ethAddress = data.ethAddress;
+    this.ipAddress = data.ipAddress;
  }
}
```

We could discuss it today and open a new issue for that, as it is way out of scope for the current issue.